Sacred logger

expectopatronum opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/williamFalcon/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs? -> I updated the documentation directly in sacred.py. Please let me know if there is documentation somewhere else
- [x] Did you write any new necessary tests? -> I wrote similar tests as for other loggers and I am testing/using it in my own project successfully. Please let me know if I should write more tests.

## What does this PR do?
Fixes #438.

## Did you have fun?
Yes, and I learned a lot :)

Borda commented on pytorch_lightning/logging/sacred.py:
```diff
@@ -0,0 +1,65 @@
+try:
+    import sacred
+except ImportError:
+    raise ImportError('Missing sacred package.')
```
pls extend the message, follow the template of the other loggers

expectopatronum replied:
I am confused, this looks like in the other loggers, e.g. https://github.com/williamFalcon/pytorch-lightning/blob/master/pytorch_lightning/logging/test_tube.py#L35

Borda replied:
ok, right, but in the logging init is the longer message...

expectopatronum replied:
I am not sure what I am looking for - could you point me to it? Thanks!

Borda replied:
ok, I am quite sure that we did it, but maybe it is still pending in other PR

expectopatronum replied:
Ah, you mean this? https://github.com/williamFalcon/pytorch-lightning/pull/648/files#diff-9fabe2a32ac055ecbb7e3a2fcab65046R46

Borda replied:
yes =)

Borda commented on pytorch_lightning/logging/sacred.py:
```diff
@@ -0,0 +1,65 @@
+except ImportError:
+    raise ImportError('Missing sacred package.')
+
+from .base import LightningLoggerBase, rank_zero_only
```
pls use absolute imports

expectopatronum replied:
I copied it from here: https://github.com/williamFalcon/pytorch-lightning/blob/master/pytorch_lightning/logging/test_tube.py#L37

Borda replied:
sure, but I think that we commented on the importing too :]

Borda commented on pytorch_lightning/logging/sacred.py:
```diff
@@ -0,0 +1,65 @@
+logger = getLogger(__name__)
+
+
+class SacredLogger(LightningLoggerBase):
```
add some description with example (sample usage)

Borda commented on pytorch_lightning/logging/sacred.py:
```diff
@@ -0,0 +1,65 @@
+
+
+class SacredLogger(LightningLoggerBase):
+    def __init__(self, database_location, database_name, experiment_name):
```
add docstring with type definition

expectopatronum commented:
@Borda It's just a draft pull request, I am aware of the things that need to be done ...

CarloLucibello commented:
If not planned already in this PR, it would be nice for lightning to intercept the Sacred observer created  by command line arguments as in
```
python script.py -F runs/     # FileObserver
python script.py -m dbruns  # MongodbObserver
```
Not sure how this would work though

expectopatronum commented:
@CarloLucibello I can think about it but I have only used MongoObserver so far. I also don't really know how the options `-F` and `-m` work. I think I'll first make it work with MongoObserver and later extend it (or someone else).

CarloLucibello commented on pytorch_lightning/logging/sacred.py:
```diff
@@ -0,0 +1,69 @@
+                url='mongodb://{ip}:{port}'.format(**mongodb_settings),
+                db_name='{db}'.format(**mongodb_settings),
+            )
+        )
```
I think the best way to do this is to have the user create the sacred observer(s) on his own and avoid creating an additional observer in this constructor. This way, we support all observers supported by sacred, without committing to the mongodb one.

expectopatronum replied:
Yes, makes a lot of sense. Then the user also has to take care of possible username/password for MongoDB for him/herself. At first I was only considering my usecase, but this will make it more useful/flexible.

expectopatronum commented:
I am currently a bit unsure how to implement `log_hyperparams`. As far as I can see, `sacred` takes care of more stuff than the other loggers that are currently implemented. It does not really allow logging hyperparamters/parameters, since it is designed to take care of that itself. I see a few options:
1) Using [sacred's config](https://sacred.readthedocs.io/en/stable/configuration.html) and it will take care of logging those parameters itself (but then I probably can't make use of pytorch lightnings grid search anymore)
2) Use lightnings hparams, log each individually using [`log_scalar`](https://sacred.readthedocs.io/en/stable/apidoc.html#sacred.Experiment.log_scalar) which is not the intended use of this method.
3) Use lightning hparams, use some other functionality of sacred to log them, e.g. using `observer.queued_event` like [here](python train_midi_guitar_piano_classifier.py --experiment_name test --batch_size 16)

Maybe I am just making things too complicated but I'd appreciate it if someone had an idea :sweat_smile: 

I found some resources that might go in the right direction (maybe we can use lightning's hparams and pass them to `run()`): 
* https://gab41.lab41.org/effectively-running-thousands-of-experiments-hyperopt-with-sacred-dfa53b50f1ec
* https://github.com/yuvalatzmon/SACRED_HyperOpt_v2/

expectopatronum commented:
Alright, after I finish implementing the test cases, I think this is ready for review. I have still a few questions left, maybe someone can answer them in the meantime:

* Do I need to add a link in README.md?
* Do I need to add somewhere else besides in `sacred.py`?
* I always get "The Travis CI build failed" - the only error I can see in the logs is "ImportError: cannot import name 'PILLOW_VERSION'". This seems to be [some issue with the installation of Pillow](https://github.com/python-pillow/Pillow/issues/4130) and it seems other current pull requests are also affected. Can/will this be fixed?

Borda commented:
https://github.com/williamFalcon/pytorch-lightning/pull/661#issuecomment-570795219

expectopatronum commented:
> [#661 (comment)](https://github.com/williamFalcon/pytorch-lightning/pull/661#issuecomment-570795219)

Ah, I am very sorry! Must have made some mistake, because I was actually searching for an issue before commenting, but I couldn't find it.

expectopatronum commented:
This pull request is ready imo, but Travis disappeared? I'd like to see if the tests run here before making it "Ready for review".

Borda commented:
@expectopatronum can you move it from Draft to normal PR status so we can see CI build?

Borda commented on tests/test_logging.py:
```diff
@@ -357,8 +357,7 @@ def run_fct():
 
-
-    result = ex.run(run_fct) # TODO: will this return the result?
+    result = ex.run(run_fct)  # TODO: will this return the result?
```
you should tell us as you are adding this logger :]

williamFalcon commented:
@expectopatronum awesome work! @Borda is this ready?

expectopatronum commented:
@williamFalcon Last time I checked the two tests were failing but I couldn't figure out (didn't have time) why. But it's probably a problem with my tests and not the code itself, because I am using it in my project and everything I expect, works. I don't see Travis now anymore, so I don't know what the status of the tests is.

Borda commented:
@williamFalcon rebase is needed... and it seems that our master is failing so I would fix the master first...

expectopatronum commented:
Hi,
at this point I don't know how to proceed. I fixed the conflicts with the base branch as suggested by @Borda. You also mentioned a rebase, is this still necessary after fixing those issues? If yes, please let me know how this should be done. 

My test case is working now locally, but Travis still fails with an error related to the TensorboardLogger. 
![tensorboardlogger](https://user-images.githubusercontent.com/2265475/72425730-a0bc5b00-3788-11ea-99b9-a56b7560e877.png)

I previously implemented two test functions:
* `test_sacred_logger` (which is fixed now)
* `test_sacred_pickle` (doesn't work and I removed it for now)

The reason for removing it is that I saw that the [wandb logger only tests very, very basic stuff](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/test_logging.py#L195).

Maybe you can clarify what functionality needs to be working & tested, since not all loggers seem to be tested the same way.

I am sorry for taking so long to finish this pull request but it's the first time I am working with Travis and at first I wasn't able to reproduce the tests locally (many more were failing bc. `pytest` didn't take care of the test requirements).

Best regards
Verena

Borda commented:
@expectopatronum  we have fixed master, could you rebase it, pls

expectopatronum commented:
@Borda I did a rebase (took me some time to figure out how that works). Is there a reason why you suggest a rebase rather than merging the changes? (from what I read while figuring out how rebase works, I got the impression merging would make more sense since I already pushed my commits to the remote).

Now it fails with the error **ERROR: torchvision 0.4.2 has requirement torch==1.3.1, but you'll have torch 1.4.0 which is incompatible.** I guess this is something that has to be fixed in your master?

Borda commented:
If you do just merge, sometimes the PR diff shows master comits as part of this PR so it is not clear what is the new addition...
Another option is creating new brach from master with the same name, do cherry-picking from the past branch and push it force.. This usually has conflicts to resolve but it is harder to explain =)
Let me see, yesterday the Master was passing with torch 1.4 I check it just I ll get to computer...

expectopatronum commented:
> If you do just merge, sometimes the PR diff shows master comits as part of this PR so it is not clear what is the new addition...
Thanks for the explanation, makes sense :)

williamFalcon commented:
@expectopatronum rebase master or add  sacred to tests/requirements.txt?

Borda commented on pytorch_lightning/logging/sacred.py:
```diff
@@ -0,0 +1,78 @@
+    def __init__(self, sacred_experiment):
+        """Initialize a sacred logger.
+
+        :param sacred.experiment.Experiment sacred_experiment: Required. Experiment object with desired observers
```
use the updated docs formating 
`:param sacred_experiment (sacred.experiment.Experiment): ... `

Borda commented on tests/test_logging.py:
```diff
@@ -386,3 +386,45 @@ def version(self):
+    """Verify that basic functionality of sacred logger works."""
+    tutils.reset_seed()
+
+    try:
```
this should not be here, the logger shall be tested

ozen commented on pytorch_lightning/logging/sacred.py:
```diff
@@ -0,0 +1,78 @@
+        return self._run_id
+
+    @rank_zero_only
+    def log_hyperparams(self, params):
```
I just wanted to share my workflow for your consideration. I use Sacred for general configuration, then Ax for hyperparameter optimization. There is one Sacred experiment and one logger during the whole process, but trainer.fit() is called many times by Ax with different hyperparameters. Logging these here somehow may be beneficial.

Borda commented:
by #767 we will do all logger imports at the beginning of the test suit...

williamFalcon commented:
@expectopatronum hi! mind rebasing and getting the tests to pass? otherwise we should close this PR until it's ready to be worked on

expectopatronum commented:
Hi, I am sorry, I don't have the time right now and I'm also not using pytorch lightning for my experiments anymore. I hoped to finish this two weeks ago but there are too many changes for me to keep track.

sheiksadique commented:
Is there still a chance if this PR making it through?

williamFalcon commented:
mind adding it to bolts?
